### PR TITLE
initial_implementation_01

### PR DESCRIPTION
**Pull request by akeaveny**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,31 +1,24 @@
-# Prerequisites
```
Comment (wmmc88) [inline .gitignore:1]: Why did u delete this?

File: LICENSE [hunk 2]
```diff
@@ -1,21 +0,0 @@
-MIT License
```
Comment (wmmc88) [inline LICENSE:1]: Why did u delete this?

File: build/CMakeCache.txt [hunk 3]
```diff
@@ -0,0 +1,330 @@
+# This is the CMakeCache file.
```
Comment (wmmc88) [inline build/CMakeCache.txt:1]: The build folder should be git ignored

**Comment by wmmc88** [comment]
When code isn't ready for merge, please make it a draft pr: https://github.blog/2019-02-14-introducing-draft-pull-requests/

File: include/rawcan_comm.hpp [hunk 4]
```diff
@@ -0,0 +1,83 @@
+#pragma once
+
+#include <linux/can.h>
+
+namespace roboteq { 
+
+class rawcan_comm : public roboteq {
+
+rawcan_comm(canid_t _roboteq_can_id, std::string _ifname);
```
Comment (wmmc88) [inline include/rawcan_comm.hpp:9]: Do we need to specifiy a can bitrate somewhere?

File: include/rawcan_comm.hpp [hunk 5]
```diff
@@ -0,0 +1,83 @@
+#pragma once
+
+#include <linux/can.h>
+
+namespace roboteq { 
+
+class rawcan_comm : public roboteq {
```
Comment (wmmc88) [inline include/rawcan_comm.hpp:7]: i believe we renamed `roboteq` here to `i_comm`

File: include/i_comm.hpp [hunk 6]
```diff
@@ -0,0 +1,83 @@
+#pragma once
+
```
Comment (wmmc88) [inline include/i_comm.hpp:2]: need to include i_comm.hpp to override these functions.

File: include/can_comm.hpp [hunk 7]
```diff
@@ -0,0 +1,76 @@
+#pragma once
```
Comment (wmmc88) [inline include/can_comm.hpp:1]: i think this file should be deleted? it was renamed to rawcan_comm right?

File: include/roboteq_controller.hpp [hunk 8]
```diff
@@ -0,0 +1,85 @@
+#pragma once
+
+#include <memory> 
+
```
Comment (wmmc88) [inline include/roboteq_controller.hpp:4]: need to include `i_comm.hpp`

File: include/roboteq_controller.hpp [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+int32_t ReadIndividualDigitalInputs(uint8_t _input);
+int16_t ReadAnalogInputs(uint8_t _input);
+int16_t ReadAnalogInputsConverted(uint8_t _input);
+int16_t ReadPulseInputs(uint8_t _input);
+int16_t ReadPulseInputsConverted(uint8_t _input);
+
+private :
+std::unique_ptr<i_comm> _comm;
+
+
```
Comment (wmmc88) [inline include/roboteq_controller.hpp:80]: get rid of extra spaces

File: src/rawcan_comm.cpp [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+        strcpy(ifr.ifr_name, ifname);
+        ioctl(_socket_handle, SIOCGIFINDEX, &ifr);
+
+        struct sockaddr_can addr;
+        
+        addr.can_family  = AF_CAN;
+        addr.can_ifindex = ifr.ifr_ifindex;
+
+        printf("%s at index %d\n", ifname, ifr.ifr_ifindex);
```
Comment (wmmc88) [inline src/rawcan_comm.cpp:38]: this is fine for now for debugging but be mindful about what gets printed in the final version.

File: src/rawcan_comm.cpp [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+        if(bind(_socket_handle, (struct sockaddr *)&addr, sizeof(addr)) < 0) {
+            printf("Error in socket bind");
+            throw -2;
```
Comment (wmmc88) [inline src/rawcan_comm.cpp:42]: this should get change to either a generic exception type, or better yet, we should create our own exception types for this library that inherit from the standard types. Not something that needs to happen for mvp, but maybe capture this in a issue for later.

File: src/roboteq_controller.cpp [hunk 12]
```diff
@@ -0,0 +1,13 @@
+
+namespace roboteq {
+
+void roboteq_controller::roboteq_controller(std::unique_ptr<i_comm> &&comm) : _comm(comm){
+
+}
+
+void roboteq_controller::SetParameters(int16_t _velocity, long _accel, long _decel, uint8_t _channel) {
+
```
Comment (wmmc88) [inline src/roboteq_controller.cpp:8]: nit: delete extra space

File: src/roboteq_controller.cpp [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+namespace roboteq {
+
+void roboteq_controller::roboteq_controller(std::unique_ptr<i_comm> &&comm) : _comm(comm){
```
Comment (wmmc88) [inline src/roboteq_controller.cpp:3]: my recollection of move semantics is not the best but i'm pretty sure this is how this works. You pass in a rvalue reference to this constructor using `std::move` on the lvalue as seen in the `example_main.cpp` and then when u pass `comm`to `_comm` in the initializer list, `comm` should be move assigned to `_comm`. if you have any trouble with these lines, lemme know.

**Review by wmmc88 — CHANGES_REQUESTED** [review]

File: include/rawcan_comm.hpp (hunk 5 above)
Comment (akeaveny) [reply include/rawcan_comm.hpp:7]: changed to i_comm

**Comment by wmmc88** [comment]
closing this to track in #7 instead

**Pull request closed without merging**